modules/core/src/async.cpp

```diff
@@ -202,7 +202,7 @@ struct AsyncArray::Impl
     {
         if (future_is_returned && refcount_future == 0)
             CV_Error(Error::StsError, "Associated AsyncArray has been destroyed");
-#ifdef CV_CXX11
+#if defined(CV_CXX11) && !defined(OPENCV_DISABLE_THREAD_SUPPORT)
         std::unique_lock<std::mutex> lock(mtx);
 #else
         cv::AutoLock lock(mtx);
```

Comment:
"#else" non C++11 branch is candidate for removal.
This patch forces to keep this.

IMHO, we need dedicated implementation for `AsyncArray` for case without threads (or even forbid using of this).

Reply:
I am happy to provide an implementation of `AsyncArray` that does not use threads. I'll give it a go, but I might need some guidance on this.

Just to make sure I am going in the right direction. You want me to have a big `#ifundef / #else / #endif` switch where the `#else` branch exposes the curret implementation (threaded) of `AsyncArray`, while the `#ifundef OPENCV_DISABLE_THREAD_SUPPORT` will implement AsyncArray without using threads (ro shoudl it be just stubs?).

Also, what the alternative...

> even forbid using of this

 ... consist of? Disabling the use of AsyncArray in all places where it is used?

Reply:
Stub with exceptions should work for now (there are no many places which uses AsyncArray)

Reply:
I started providing stubs for the API of `cv::AsyncArray`, and it turned out I also need to provide stubs for `cv::AsyncPromise`. The approach I followed was to remove the `AsyncArray::Impl` definition and uses. This however caused writing many `#ifundef / #else / #endif` cases, which doesn't seem to fit very well with the goal of miniming such cases.

Would it be OK if instead of providing stubs for `cv::AsyncArray` and `cv::AsyncPromise`, I would be providing  a stub class `AsyncArray::Impl` that would expose stubs only for the methods used by `cv::AsyncArray` and `cv::AsyncPromise`? This solution would still provide useful exception message, and would require just a couple of `#ifundef / #else / #endif` cases.